Implement tensor.refine_names

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #25843 Implement tensor.align_as(other), change tensor.align_to(names)
* **#25842 Implement tensor.refine_names**

`tensor.refine_names(*names)` takes `tensor` and attempts to name its
dimensions `names` out-of-place. If a dimension `i` already had a name,
then it cannot be changed (so tensor.names[i] must equal names[i]);
if the original dimension did not have a name, then the new name
(names[i]) can be anything.

`tensor.refine_names(*names)` also accepts a glob '*' that greedily selects
names from `tensor`. Here are some examples:

- `Tensor[None].refine_names('N') -> Tensor[N]`
- `Tensor[N].refine_names('N') -> Tensor[N]`
- `Tensor[N].refine_names('D') -> Error!`
- `Tensor[N].refine_names(None) -> Error!`
- `Tensor[None, None].refine_names('*', D) -> Tensor[None, D]`

Test Plan:
- new tests [namedtensor ci]

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25842

Differential Revision: [D17255548](https://our.internmc.facebook.com/intern/diff/D17255548)

zou3519 committed:
Implement tensor.as_named

`tensor.as_named(*names)` takes `tensor` and attempts to name its
dimensions `names` out-of-place. If a dimension `i` already had a name,
then it cannot be changed (so tensor.names[i] must equal names[i]);
if the original dimension did not have a name, then the new name
(names[i]) can be anything.

`tensor.as_named(*names)` also accepts a glob '*' that greedily selects
names from `tensor`. Here are some examples:

- `Tensor[None].as_named('N') -> Tensor[N]`
- `Tensor[N].as_named('N') -> Tensor[N]`
- `Tensor[N].as_named('D') -> Error!`
- `Tensor[N].as_named(None) -> Error!`
- `Tensor[None, None].as_named('*', D) -> Tensor[None, D]`

Test Plan:
- new tests [namedtensor ci]

nairbv commented on aten/src/ATen/native/NamedTensor.cpp:
```diff
@@ -89,6 +89,38 @@ static std::vector<int64_t> aligned_size(
+    }
+    if (out_name.is_wildcard()) {
+      TORCH_CHECK(false,
+          "as_named: cannot coerse Tensor", self_names, " to Tensor", names,
```
what's the use-case for a function that must not allow re-naming vs just having a rename function?

zou3519 replied:
Users can use `as_named` to assert names on a tensor. If they want to check that the names are `NCHW` they can use `tensor.as_named('N', 'C', 'H', 'W')`.

For example,

```
def conv(input):
    input = input.renamed('N', 'C', H', 'W')
    # do things
```
`renamed` is not the right function here because the input might have had names `('N', 'H', 'W', 'C')`.

nairbv replied:
Since it's a new function we probably need docs for it (e.g. add_docstr in _torch_docs.py).

It's a bit confusing to me if as_named is meant to be an assertion more than a rename.

If you feel confident this is important to have that's fine, but in general let's bias towards adding functions / name-support as minimally as possible. Post-release it will be easier to add more than to remove them or continue to support them if we find any that aren't necessary.

zou3519 replied:
Named tensor docs will come later in a batch. I'm trying to get a demo up and running asap.

`as_named` is meant to coerce unnamed, partially named, and fully named tensors to having the desired set of names positionally. As a side effect, one can use it to assert on named tensors. I am confident we need this functionality (it's been around since the beginning of the design phase) but it's possible that we will end up splitting it into two functions or renaming it.

zou3519 replied:
Perhaps a better name would be `refine_names`?

zou3519 committed:
Update on "Implement tensor.as_named"


`tensor.as_named(*names)` takes `tensor` and attempts to name its
dimensions `names` out-of-place. If a dimension `i` already had a name,
then it cannot be changed (so tensor.names[i] must equal names[i]);
if the original dimension did not have a name, then the new name
(names[i]) can be anything.

`tensor.as_named(*names)` also accepts a glob '*' that greedily selects
names from `tensor`. Here are some examples:

- `Tensor[None].as_named('N') -> Tensor[N]`
- `Tensor[N].as_named('N') -> Tensor[N]`
- `Tensor[N].as_named('D') -> Error!`
- `Tensor[N].as_named(None) -> Error!`
- `Tensor[None, None].as_named('*', D) -> Tensor[None, D]`

Test Plan:
- new tests [namedtensor ci]

Differential Revision: [D17255548](https://our.internmc.facebook.com/intern/diff/D17255548)

zou3519 committed:
Update on "Implement tensor.refine_names"


`tensor.refine_names(*names)` takes `tensor` and attempts to name its
dimensions `names` out-of-place. If a dimension `i` already had a name,
then it cannot be changed (so tensor.names[i] must equal names[i]);
if the original dimension did not have a name, then the new name
(names[i]) can be anything.

`tensor.refine_names(*names)` also accepts a glob '*' that greedily selects
names from `tensor`. Here are some examples:

- `Tensor[None].refine_names('N') -> Tensor[N]`
- `Tensor[N].refine_names('N') -> Tensor[N]`
- `Tensor[N].refine_names('D') -> Error!`
- `Tensor[N].refine_names(None) -> Error!`
- `Tensor[None, None].refine_names('*', D) -> Tensor[None, D]`

Test Plan:
- new tests [namedtensor ci]

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25842

Differential Revision: [D17255548](https://our.internmc.facebook.com/intern/diff/D17255548)
